### PR TITLE
feat: KV cache precomputation for static prompt prefix

### PR DESCRIPTION
**Pull request by paulnsorensen**
## Summary
- **MLX backend**: Precompute and reuse KV cache for the static prompt prefix (system prompt + rule template). Per-request `deepcopy` isolation, MD5-keyed cache dict, BPE boundary validation on first warm.
- **GGUF backend**: Enable `LlamaRAMCache` (2 GB capacity) for automatic longest-prefix token matching — zero protocol changes needed.
- **Protocol layer**: Add `prefix_len` field to `ClassifyRequest`, `Rule.split_prompt()` to separate static prefix from dynamic user text, and `CachedBackend`/`CachedLogprobBackend` runtime-checkable Protocols for type-safe dispatch.
- **Daemon routing**: `isinstance`-based elif chain dispatches to cached or uncached inference paths based on backend capabilities and `prefix_len > 0`.

## Key design decisions
- Runner owns the split (via `Rule.split_prompt`), daemon stays dumb — receives `prefix_len` alongside the prompt
- Chat template splitting uses SPLIT_MARKER technique with ChatML fallback
- BPE boundary mismatch logs a warning but doesn't block (graceful degradation)

## Test plan
- [x] 12 tests for `Rule.split_prompt` (prefix_len correctness, sanitization, back-truncation, edge cases)
- [x] 6 tests for daemon cached inference routing (all dispatch paths + fallbacks)
- [x] ~15 tests for MLX cached methods (warm, reuse, deepcopy isolation, logprobs, EOS)
- [x] `just check` passes (lint + typecheck + 439 tests)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: vaudeville/core/rules.py [hunk 1]
```diff
@@ -92,6 +92,20 @@ def format_prompt(self, text: str, context: str = "") -> str:
...
+
+        prefix_len is the character index where the static prefix ends
+        and the variable {text} content begins.
+        """
+        safe_text = _sanitize_input(back_truncate(text))
+        safe_context = _sanitize_input(context) if context else ""
+        prompt_with_context = self.prompt.replace("{context}", safe_context)
+
+        before, _, after = prompt_with_context.partition("{text}")
+        full_prompt = before + safe_text + after
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/core/rules.py:106]: Intentional. Every existing rule has exactly one `{text}` — a classification prompt interpolating user input in multiple places makes no semantic sense. `format_prompt` uses `.replace()` which also replaces all occurrences, but the real-world count is always one. Adding a validator for a constraint that's already enforced by the rule authoring convention would be YAGNI.

File: vaudeville/server/daemon.py [hunk 2]
```diff
@@ -71,10 +89,11 @@ def handle_request(
     try:
         request = json.loads(data.decode().strip())
         prompt = request.get("prompt", "")
+        prefix_len = request.get("prefix_len", 0)
 
-        logger.debug("prompt=%d chars", len(prompt))
+        logger.debug("prompt=%d chars prefix_len=%d", len(prompt), prefix_len)
         t0 = time.monotonic()
-        result = _run_inference(backend, prompt)
+        result = _run_inference(backend, prompt, prefix_len)
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/daemon.py:96]: The daemon wraps the entire `handle_request` body in `try/except Exception` (lines 113-115), returning fail-open on any error — including a non-int `prefix_len`. The only caller is `VaudevilleClient`, which always sends an int. This is a localhost Unix socket, not a public API. Extra coercion would be dead code.

File: vaudeville/server/mlx_backend.py [hunk 3]
```diff
@@ -54,32 +57,178 @@ def classify_with_logprobs(
...
+        Logs a warning if BPE boundary effects cause divergence.
+        """
+        formatted_prefix = self._format_prefix(user_prefix)
+        prefix_key = hashlib.md5(formatted_prefix.encode()).hexdigest()
+        base_cache = self._prefix_caches.get(prefix_key)
+        if base_cache is None:
+            self._validate_token_boundary(formatted_prefix, user_prefix)
+            base_cache = self._warm_prefix(formatted_prefix)
+            self._prefix_caches[prefix_key] = base_cache
+        return copy.deepcopy(base_cache)
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/mlx_backend.py:117]: Valid concern — the cache was unbounded. Fixed in a4a77ef: `_prefix_caches` is now an `OrderedDict` with LRU eviction capped at `MAX_PREFIX_CACHES = 16`.

File: vaudeville/server/mlx_backend.py [hunk 4]
```diff
@@ -54,32 +57,178 @@ def classify_with_logprobs(
...
+                [
+                    {"role": "system", "content": SYSTEM_PROMPT},
+                    {"role": "user", "content": "SPLIT_MARKER"},
+                ],
+                tokenize=False,
+                add_generation_prompt=True,
+            )
+            if "SPLIT_MARKER" in full:
+                opening, _, closing = full.partition("SPLIT_MARKER")
+                return opening, closing
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply vaudeville/server/mlx_backend.py]: Fixed in ab0a209. Cached the `(opening, closing)` tuple as `self._chat_template_parts` on first call. The template is static after `__init__`, so subsequent calls to `_format_prefix` and `_format_suffix` hit the in-memory cache instead of re-invoking `apply_chat_template`.

File: tests/test_gguf_backend.py [hunk 5]
```diff
@@ -30,6 +30,7 @@ def test_classify_returns_string(self) -> None:
             {
                 "huggingface_hub": MagicMock(hf_hub_download=mock_hub),
                 "llama_cpp": MagicMock(Llama=mock_lm_cls),
+                "llama_cpp.llama_cache": MagicMock(),
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:33]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 6]
```diff
@@ -49,6 +50,7 @@ def test_classify_passes_max_tokens(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:53]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 7]
```diff
@@ -69,6 +71,7 @@ def test_classify_includes_system_prompt(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:74]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 8]
```diff
@@ -109,6 +112,7 @@ def test_classify_with_logprobs_includes_system_prompt(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:115]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 9]
```diff
@@ -133,6 +137,7 @@ def test_classify_passes_repeat_penalty(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:140]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 10]
```diff
@@ -170,6 +175,7 @@ def test_classify_with_logprobs_passes_repeat_penalty(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:178]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 11]
```diff
@@ -194,6 +200,7 @@ def test_classify_passes_grammar(self) -> None:
                 "llama_cpp": MagicMock(
                     Llama=mock_lm_cls, LlamaGrammar=mock_grammar_cls
                 ),
+                "llama_cpp.llama_cache": MagicMock(),
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:203]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 12]
```diff
@@ -241,6 +248,7 @@ def test_classify_with_logprobs_omits_grammar(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:251]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_gguf_backend.py [hunk 13]
```diff
@@ -265,6 +273,7 @@ def test_grammar_is_cached_across_calls(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_gguf_backend.py:276]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_backends.py [hunk 14]
```diff
@@ -59,6 +60,7 @@ def test_classify_passes_max_tokens(self) -> None:
             {
                 "huggingface_hub": MagicMock(hf_hub_download=mock_hub),
                 "llama_cpp": MagicMock(Llama=mock_lm_cls),
+                "llama_cpp.llama_cache": MagicMock(),
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_backends.py:63]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_backends.py [hunk 15]
```diff
@@ -183,7 +185,11 @@ def test_calls_download_and_inference(self) -> None:
         mock_llama_cpp = MagicMock(Llama=mock_llama_cls)
         with patch.dict(
             "sys.modules",
-            {"huggingface_hub": mock_hub, "llama_cpp": mock_llama_cpp},
+            {
+                "huggingface_hub": mock_hub,
+                "llama_cpp": mock_llama_cpp,
+                "llama_cpp.llama_cache": MagicMock(),
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_backends.py:191]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.

File: tests/test_backends.py [hunk 16]
```diff
@@ -40,6 +40,7 @@ def test_classify_returns_string(self) -> None:
             {
                 "huggingface_hub": MagicMock(hf_hub_download=mock_hub),
                 "llama_cpp": MagicMock(Llama=mock_lm_cls),
+                "llama_cpp.llama_cache": MagicMock(),
```
[the remark this answers is not in the document]
Comment (paulnsorensen) [reply tests/test_backends.py:43]: This is not a bug. `MagicMock()` auto-creates attributes on access, so `from llama_cpp.llama_cache import LlamaRAMCache` succeeds — it resolves to a MagicMock attribute rather than raising `ImportError`. All 439 tests pass. This is standard Python mocking behavior for patching module graphs.
